app/lib/authorizer.rb

```diff
@@ -0,0 +1,144 @@
+# This class defines authorization rules.
+# If you want student data, ask through this class.
+# Ideally, start with using `authorized`.
+# Drop down to calling individual methods if you need to.
+#
+# This class should be fully tested, and all changes should be reviewed.
+#
+# Controller usage:
+#   student = authorized { Student.find(params[:id]) }
+#   homeroom = authorized { Homeroom.(params[:id]) }
+#   students = authorized { homeroom.students }
+#
+# Standalone usage:
+#   authorizer = Authorizer.new(educator)
+#   student = authorizer.authorized { Student.find(params[:id]) }
+#   homeroom = authorizer.authorized { Homeroom.(params[:id]) }
+#   students = authorizer.authorized { homeroom.students }
+#
+# Instead of:
+#   Homeroom.find(id).students # no authorization check
+#   Section.find(id).students # no authorization check
+#   current_educator.homeroom.students # doesn't work for all roles because
+#     some roles don't have homeroom associations (e.g. if they're school-wide)
+#   current_educator.school.students # doesn't work for all roles because
+#     some roles don't have school associations (e.g. if they're district-wide)
+
+#
+class Authorizer
+  # These fields are required on the `Student` model for
+  # authorization checks
+  def self.student_fields_for_authorization
+    [
+      :id,
+      :program_assigned,
+      :limited_english_proficiency,
+      :school_id,
+      :grade
+    ]
+  end
+
+  def initialize(educator)
+    @educator = educator
+    @authorized_dispatcher = AuthorizedDispatcher.new(self)
+  end
+
+  # Usage:
+  #  student = authorized { Student.find(params[:id]) }
+  def authorized(&block)
+    @authorized_dispatcher.dispatch(&block)
+  end
+
+  # This method is the source of truth for whether an educator is authorized to view information about a particular
+  # student.
+  def is_authorized_for_student?(student)
+    begin
+      return true if @educator.districtwide_access?
+
+      return false if @educator.restricted_to_sped_students && !(student.program_assigned.in? ['Sp Ed', 'SEIP'])
+      return false if @educator.restricted_to_english_language_learners && student.limited_english_proficiency == 'Fluent'
+      return false if @educator.school_id.present? && student.school_id.present? && @educator.school_id != student.school_id
+
+      return true if @educator.schoolwide_access? || @educator.admin? # Schoolwide admin
+      return true if @educator.has_access_to_grade_levels? && student.grade.in?(@educator.grade_level_access) # Grade level access
+      return true if student.in?(@educator.students) # Homeroom level access
+      return true if student.in?(@educator.section_students) # Section level access
+    rescue ActiveModel::MissingAttributeError => err
```

Comment:
I have a question about the `MissingAttributeError` rescue here!  

When I look at the code above, I don't see how any of it could trigger a `MissingAttributeError` error. Here's how I'd imagine a line like `@educator.restricted_to_sped_students` would work:

```
# @educator exists
@educator.restricted_to_sped_students
 => false 
 => true

# does not exist
nil.restricted_to_sped_students
=> NoMethodError: undefined method `restricted_to_sped_students' for nil:NilClass
```

I imagine the same for the `student` variable. 

So where might we see `MissingAttributeError` being raised in the code?

Reply:
OK so now thanks to our other thread of conversation about this method I'm looking at https://apidock.com/rails/v4.0.2/ActiveRecord/QueryMethods/select:

```
Accessing attributes of an object that do not have fields retrieved by a select will throw ActiveModel::MissingAttributeError:
```

Reply:
But what's confusing me here is that `is_authorized_for_student?` is down at the individual student level, whereas `select` is called on an entire class or an active record relation, like this:

```
2.3.0 :001 > Student.select(:local_id)
2.3.0 :002 > Student.all.select(:local_id)
```

So it would seem like we're rescuing at the wrong level here.

Reply:
@alexsoble I think we talked about this the other day, but when `select` is used on the class or relation like that, it returns active record models that only have some of the attributes of the regular ActiveRecord model.  This is an edge case, but came in trying to update the `students_controller#lasids` method to use the `authorized` API here.  That code was trying to grab all students but only a field or two, and so this would raise to say "hey you have to add some more fields so that the authorization checks can work."